nat/traversal/nat_proxy.go

```diff
@@ -21,22 +21,150 @@ import (
+// Read from listener socket and write to remoteConn
+// Read from remoteConn and write to listener socket
+func (np *NATProxy) consumerHandOff(consumerPort int, remoteConn *net.UDPConn) {
+	log.Info(logPrefix, "Inside consumer NAT proxy")
+
+	// reinit a stop channel, since it might be closed by previous connect
+	np.stop = make(chan struct{})
+
+	// outIP, err := np.ipResolver.GetOutboundIP()
+	laddr, err := net.ResolveUDPAddr("udp4", fmt.Sprintf("%s:%d", "127.0.0.1", consumerPort+1))
+	if err != nil {
+		log.Error(logPrefix, "failed to get local address for consumer NAT proxy: ", err)
+		return
+	}
+
+	remoteConn.SetReadBuffer(bufferLen)
+	remoteConn.SetWriteBuffer(bufferLen)
+
+	fd, err := remoteConn.File()
+	if err != nil {
+		log.Error(logPrefix, "failed to fetch fd from: ", remoteConn)
+	}
+	defer fd.Close()
+
+	log.Info(logPrefix, "protecting socket: ", int(fd.Fd()))
+
+	np.socketProtect(int(fd.Fd()))
+
+	for {
+		log.Info(logPrefix, "waiting connect from openvpn3 client process")
+		// If for some reason consumer disconnects, new connection will be from different port
+		proxyConn, err := net.ListenUDP("udp4", laddr)
+		if err != nil {
+			log.Errorf("%sfailed to listen for consumer proxy on: %v, %v", logPrefix, laddr, err)
+			return
+		}
+
+		select {
+		case <-np.stop:
+			log.Info(logPrefix, "Stopping NATProxy handOff loop")
+			proxyConn.Close()
+			remoteConn.Close()
+			return
+		default:
+			/*
+				proxyFD, err := proxyConn.File()
+				if err != nil {
+					log.Error(logPrefix, "failed to fetch fd from: ", proxyConn)
+				}
+
+				log.Info(logPrefix, "protecting local socket: ", int(proxyFD.Fd()))
+				np.socketProtect(int(proxyFD.Fd()))
+			*/
+			proxyConn.SetReadBuffer(bufferLen)
+			proxyConn.SetWriteBuffer(bufferLen)
+
+			np.masterLoop(proxyConn, remoteConn)
+
+			proxyConn.Close()
+			//			proxyFD.Close()
+		}
+	}
+}
+
+func (np *NATProxy) masterLoop(conn *net.UDPConn, remoteConn *net.UDPConn) {
+	log.Info(logPrefix, "start copying stream from consumer NAT proxy to remote remoteConn")
+	for {
+		select {
+		case <-np.stop:
+			log.Info(logPrefix, "Stopping NATProxy masterLoop")
+			return
+		default:
+			var buf [bufferLen]byte
+			n, addr, err := conn.ReadFromUDP(buf[0:])
+			if err != nil {
+				log.Errorf("%sFailed	 to read local process: %s cause: %s", logPrefix, conn.LocalAddr().String(), err)
+				return
+			}
+			if n > 0 {
+				_, err := remoteConn.Write(buf[:n])
+				if err != nil {
+					log.Errorf("%sFailed to write remote peer: %s cause: %s", logPrefix, remoteConn.RemoteAddr().String(), err)
+					return
+				}
+				if np.addrLast != addr {
+					np.addrLast = addr
+					go np.readWriteToAddr(remoteConn, conn, addr)
+				}
+			}
+		}
+	}
+}
+
+func (np *NATProxy) readWriteToAddr(conn *net.UDPConn, remoteConn *net.UDPConn, addr *net.UDPAddr) {
+	for {
+		select {
+		case <-np.stop:
+			log.Info(logPrefix, "Stopping NATProxy readWriteToAddr loop")
+			return
+		default:
```

Comment:
We can make nesting level a bit smaller with empty `default` block since the first case ends with return:
```suggestion
		default:
	}
```